### PR TITLE
Refactor to use local copy of heap_4 for efr32 boards

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Problem
* Suffering from Heap fragmentation during PASE and/or CASE session establishement which causes the session establishement to fail
* Tracking of heap memory allocation does not track every single memory allocation

#### Change overview
* Refactor efr32 code to use a copy of heap_4 since it's more effective in avoiding heap fragmentation.
* Forces all heap memory allocations to use the memory allocation functions in the heap_4 copy by adding build flags to change default heap memory allocation functions to wrapper functions.

The changes are to go from two heaps that are not overlapping to one single heap managed by FreeRTOS. The build arguments and wrapper functions force all memory allocations to go into the FreeRTOS heap. With these changes, all memory allocationa go trough the heap_4_silabs memory allocation functions so we can have a better tracking of free's, allocation's and memory fragmentation. It also allows us to taylor the tracking to our need in the futur.

#### Testing
* Built all the examples supported for efr32 on all supported boards
* Tested changes with one example on a board

### CODE REVIEW

File: src/platform/EFR32/CHIPMem-Platform.cpp [hunk 1]
```diff
@@ -118,31 +116,33 @@ void * MemoryAlloc(size_t size, bool isLongTermAlloc)
...
     return ptr;
 }
 
 void * MemoryRealloc(void * p, size_t size)
 {
     VERIFY_INITIALIZED();
-    return sl_realloc(p, size);
+
+    MemoryFree(p);
+    p = MemoryAlloc(size);
```
Comment (msandstedt) [inline src/platform/EFR32/CHIPMem-Platform.cpp:137]: This most definitely will not work.

**Review by msandstedt — CHANGES_REQUESTED** [review]
Noted inline.  `MemoryRealloc` is broken.

**Review by jmartinez-silabs — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: examples/platform/efr32/heap_4_silabs.c [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+            }
+            else // New size is a different from current size
+            {
+                resized_ptr = pvPortCalloc(1, size);
+                if (resized_ptr != NULL)
+                {
+                    int smallest_size = size < current_size ? size : current_size;
+                    memcpy(resized_ptr, pv, smallest_size);
+                }
+                vPortFree(pv);
```
Comment (bzbarsky-apple) [inline examples/platform/efr32/heap_4_silabs.c:320]: Shouldn't this `vPortFree` be inside the `if (resized_ptr != NULL)` block?  The usual semantics for realloc is that if it fails the input pointer remains valid.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/platform/efr32/heap_4_silabs.c (hunk 2 above)
Comment (mkardous-silabs) [reply examples/platform/efr32/heap_4_silabs.c]: Yes it should be, I pushed the change.

**Comment by mspang** [comment]
Can you explain the problem in more detail?

**Comment by jmartinez-silabs** [comment]
@mspang We saw previously that we were suffering from heap fragmentation during PASE and/or CASE session establishment. We increased the heap in past commits to prevent failures where we didn't have any more block big enough for the required allocation. 
Heap 4 is a bit better at reducing heap fragmentation and also already has better functionalities to track the heap usage/stats. We will also be able to easily add to it. 

The PR also add some wrappers for allocations that were not previously tracked.

**Comment by mspang** [comment]
> @mspang We saw previously that we were suffering from heap fragmentation during PASE and/or CASE session establishment. We increased the heap in past commits to prevent failures where we didn't have any more block big enough for the required allocation.
> Heap 4 is a bit better at reducing heap fragmentation and also already has better functionalities to track the heap usage/stats. We will also be able to easily add to it.

That's a great description, can you put it in the PR's description?

> 
> The PR also add some wrappers for allocations that were not previously tracked.

What are you referring to by "tracked"?

**Comment by mkardous-silabs** [comment]
@mspang We added build flags and wrapper functions to change default heap allocation functions as to garantee all heap memory allocations are beeing tracked.

**Comment by mspang** [comment]
> @mspang We added build flags and wrapper functions to change default heap allocation functions as to garantee all heap memory allocations are beeing tracked.

It's the verb "track" that I think should be clarified.. what does it mean for an allocation to be not "tracked" ? The purpose of malloc is to allocate the memory.

**Comment by mspang** [comment]
> > @mspang We added build flags and wrapper functions to change default heap allocation functions as to garantee all heap memory allocations are beeing tracked.
> 
> It's the verb "track" that I think should be clarified.. what does it mean for an allocation to be not "tracked" ? The purpose of malloc is to allocate the memory.

Here's some specific guesses I have about it:

- We're using two heaps and they use non-overlapping memory regions not used for any purpose (this is perfectly safe but perhaps not desirable)
- We're using two heaps and the regions overlap (this one will cause severe breakage)
- Allocations all end up in one heap but some extra accounting done by a wrapper is wrong.
- Allocations all end up in one heap but not via the desired sequence of API calls. This is a superficial issue.

**Comment by mkardous-silabs** [comment]
@mspang 
The changes for the tracking aspect is to go from two heaps that are not overlapping to one single heap managed by FreeRTOS. The build arguments and wrapper functions force all memory allocation to go into the FreeRTOS heap.
Whit this change, all memory allocation function go trough the heap_4_silabs memory allocation function so we can have a better tracking of free's, allocation's and memory fragmentation. It also allows us to taylor the tracking to our need in the futur.

**Comment by mspang** [comment]
> @mspang
> The changes for the tracking aspect is to go from two heaps that are not overlapping to one single heap managed by FreeRTOS. The build arguments and wrapper functions force all memory allocation to go into the FreeRTOS heap.
> Whit this change, all memory allocation function go trough the heap_4_silabs memory allocation function so we can have a better tracking of free's, allocation's and memory fragmentation. It also allows us to taylor the tracking to our need in the futur.
> 
> I'll add the information to the PR description at the same time.

Perfect, thank you.

**Review by mspang — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**